Assets/Scripts/Common/AState.cs

```diff
@@ -0,0 +1,42 @@
+/// <summary>
+/// State abstract base class; It provides a simple way to implement new states
+/// by giving a default implementation for all methods of the IState interface;
+/// Only used methods should be overriden by child classes
+/// </summary>
+abstract public class AState<ContextT> : IState
+  where ContextT : AStateContext
+{
+    protected ContextT context;
+
+    public void SetContext(AStateContext context)
+    {
+        this.context = context as ContextT;
+    }
+
+    virtual public void Enter()
+    {
+    }
+
+    virtual public void Update()
+    {
+    }
+
+    virtual public void FixedUpdate()
+    {
+    }
+
+    virtual public void LateUpdate()
+    {
+    }
+
+    virtual public void Exit()
+    {
+    }
+}
+
+/// <summary>
+/// Default AState implementation, inheriting  from AStateContext
+/// </summary>
+abstract public class AState : AState<AStateContext>
```

Comment:
What is the purpose of this class?

Reply:
If you want to create a state and don't want to bother with what type of context it depends on, you can say it depends on an AStateContext instead of a specific context type (like PlayerStateContext or CameraStateContext for instance).
This class allows you to do so by default, by making your state inherit from `AState` instead of `AState<AStateContext>`: it's only a shortcut/syntactic sugar.

Example:
```csharp
class IdleState : AState<AStateContext>
{
    // do nothing
}
```
could be written
```csharp
class IdleState : AState
{
    // do nothing
}
```